File: sprint2/By lessons/lesson9code1/lesson9code1.kt
class Queue(private val n: Int) {
    private val queue = arrayOfNulls<Int>(n)
    private val maxN = n
    private var head = 0
    private var tail = 0
    private var size = 0

    fun isEmpty() = size == 0

    fun push(x: Int) {
        if (size != maxN) {
            queue[tail] = x
            tail = (tail + 1) % maxN
            size += 1
        }
    }

    fun pop(): Int? {
        if (isEmpty()) {
            return null
        }
        val x = queue[head]
        queue[head] = null
        head = (head + 1) % maxN
        size -= 1
        return x
    }
}

val q = Queue(8)
q.push(-1)
q.push(0)
q.push(11)

println(q.pop())   // -1
println(q.queue.toList())  // [None, -1, 0, 11, None, None, None, None]
println(q.size)   // 3

q.pop()
q.pop()
q.push(-8)
q.push(7)
q.push(3)
q.push(16)
println(q.queue.toList()) // [None, None, None, 11, -8, 7, 3, 16]
println(q.size) // 5
q.push(12)
println(q.queue.toList()) // [12, None, None, 11, -8, 7, 3, 16]
println(q.size) // 6

File: sprint2/By lessons/lesson9code1/lesson9code1.go
package main

import (
    "fmt"
)

type Queue struct {
    queue  []interface{}
    maxN   int
    head   int
    tail   int
    size   int
}

func NewQueue(n int) *Queue {
    return &Queue{
        queue:  make([]interface{}, n),
        maxN:   n,
        head:   0,
        tail:   0,
        size:   0,
    }
}

func (q *Queue) IsEmpty() bool {
    return q.size == 0
}

func (q *Queue) Push(x interface{}) {
    if q.size != q.maxN {
        q.queue[q.tail] = x
        q.tail = (q.tail + 1) % q.maxN
        q.size += 1
    }
}

func (q *Queue) Pop() interface{} {
    if q.IsEmpty() {
        return nil
    }
    x := q.queue[q.head]
    q.queue[q.head] = nil
    q.head = (q.head + 1) % q.maxN
    q.size -= 1
    return x
}

func main() {
    q := NewQueue(8)

    q.Pop()
    fmt.Println(q.queue) // [nil -1 0 11 nil nil nil nil]
    fmt.Println(q.size)  // 3

    q.Pop()
    q.Pop()
    q.Push(-8)
    q.Push(7)
    q.Push(3)
    q.Push(16)
    fmt.Println(q.queue) // [nil nil nil 11 -8 7 3 16]
    fmt.Println(q.size)  // 5

    q.Push(12)
    fmt.Println(q.queue) // [12 nil nil 11 -8 7 3 16]
    fmt.Println(q.size)  // 6
}
